### PR TITLE
Bookshelf Integration

### PR DESCRIPTION
**Pull request by casrom**
Complete bookshelf backend (userbook), and integrate with front-end bookshelf functions.

### CODE REVIEW

File: server/book/book.service.ts [hunk 1]
```diff
@@ -43,14 +43,15 @@ export class BookService {
...
+    let book: any = await this.bookModel.findOne({ gid: id }).exec();
     const url = `https://www.googleapis.com/books/v1/volumes/${id}`;
-    return this.httpService
-      .get<GoogleBookItem>(url)
-      .pipe(map((response) => this.mapGoogleBooks(response.data)));
+    if (!book) {
+      book = await this.httpService
+        .get<GoogleBookItem>(url)
+        .pipe(map((response) => this.mapGoogleBooks(response.data)))
+        .toPromise();
```
Comment (hans-lizihan) [inline server/book/book.service.ts:52]: i think `toPromise` is a deprecating method in rxjs.

i think the general advise here is use `firstValueFrom` 

https://rxjs.dev/api/index/function/firstValueFrom

File: server/book/book.service.ts [hunk 2]
```diff
@@ -43,14 +43,15 @@ export class BookService {
   }
 
   async findById(id: string) {
-    const book = await this.bookModel.findOne({ gid: id }).exec();
-    if (book) {
-      return book;
-    }
+    let book: any = await this.bookModel.findOne({ gid: id }).exec();
```
Comment (hans-lizihan) [inline server/book/book.service.ts:46]: and book should have `Book` type (as in the book model's Book type)

File: server/userbook/userbook.service.ts [hunk 3]
```diff
@@ -8,8 +8,9 @@ import { BookService } from '../book/book.service';
 @Injectable()
 export class UserBookService {
   constructor(
-    @InjectModel(UserBook.name) private userBookModel: Model<UserBook>,
+    @Inject(forwardRef(() => BookService))
     private bookService: BookService,
+    @InjectModel(UserBook.name) private userBookModel: Model<UserBook>,
   ) {}
 
   async findByUidAndBookshelf(user_id: string, bookshelf: BookshelfType) {
```
Comment (hans-lizihan) [inline server/userbook/userbook.service.ts:16]: and i think you should change all snake_case based `user_id` to `userId` everywhere to keep consistent

File: client/pages/BookshelfPage/BookshelfPage.tsx [hunk 4]
```diff
@@ -9,21 +9,33 @@ const SHELVES = [
...
-  { title: 'Read', url: '/bookshelves/read', key: 'read' },
-  { title: 'Favorite', url: '/bookshelves/favorite', key: 'favorite' },
 ];
 
 const URLS = SHELVES.reduce<{ [key: string]: string }>((acc, cur) => {
   acc[cur.key] = cur.url;
   return acc;
 }, {});
 
+const QUERY_KEYS = SHELVES.reduce<{ [key: string]: string }>((acc, cur) => {
```
[the remark this answers is not in the document]
Comment (hans-lizihan) [reply client/pages/BookshelfPage/BookshelfPage.tsx:34]: to mitigate this, you could use lodash `groupBy` function and `mapValues` function


https://lodash.com/docs/4.17.15#groupBy
https://lodash.com/docs/4.17.15#mapValues


```
const MAPPING = groupBy(SHELVES, 'key')
const QUERY_KEYS = mapValues(MAPPING, 'queryKey')
const URLS = mapValues(MAPPING, 'url')
```

File: client/components/BookCard.tsx [hunk 5]
```diff
@@ -41,6 +42,7 @@ export default function BookCard(props: Props) {
         >
           <Rating size="small" name="read-only" value={3} readOnly /> 20 reviews
         </Typography>
+        <BookshelfActions bookData={book} iconsOnly={true} />
```
Comment (hans-lizihan) [inline client/components/BookCard.tsx:45]:
```suggestion
        <BookshelfActions bookData={book} iconsOnly />
```

`={true}` could be emitted as per react community convention

File: client/components/BookshelfActions.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import ButtonGroup from '@mui/material/ButtonGroup';
+import FavoriteIcon from '@mui/icons-material/Favorite';
+import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder';
+import CheckCircleOutlineIcon from '@mui/icons-material/CheckCircleOutline';
+import CheckCircleIcon from '@mui/icons-material/CheckCircle';
+import BookmarkBorderIcon from '@mui/icons-material/BookmarkBorder';
+import BookmarkIcon from '@mui/icons-material/Bookmark';
+
+export interface Props {
+  bookData: Book;
```
Comment (hans-lizihan) [inline client/components/BookshelfActions.tsx:15]:
```suggestion
  book: Book;
```

bookData is a verbose and redundant prop name

File: client/components/BookshelfActions.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  orientation?: 'vertical';
+  iconsOnly?: boolean;
+}
+
+export default function BookshelfActions(props: Props) {
+  const { bookData, orientation, iconsOnly } = props;
+
+  const mutation = useMutation(
+    (payload: Book & { bookshelf?: string; favorite?: boolean }) =>
```
Comment (hans-lizihan) [inline client/components/BookshelfActions.tsx:24]: bookshelf and favorite could be optional?

File: client/components/BookshelfActions.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      fetcher.post(`/api/bookshelves`, payload),
+  );
+
+  const onButtonClick = (label: string) => {
```
Comment (hans-lizihan) [inline client/components/BookshelfActions.tsx:28]:
```suggestion
  const handleButtonClick = (label: string) => {
```

File: client/components/BookshelfActions.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        ...bookData,
+        bookshelf: label,
+      });
+    } else {
+      mutation.mutate({
+        ...bookData,
+        favorite: true,
+      });
+    }
+    // TODO: trigger book data refetch somehow
```
Comment (hans-lizihan) [inline client/components/BookshelfActions.tsx:40]: react query has some machanism for this. I think you could specify the query key in the first argument in `useMutation`. 

```
useMutation(['bookshelf'], ...)
```
```

File: client/components/BookshelfActions.tsx (hunk 7 above)
Comment (casrom) [reply client/components/BookshelfActions.tsx:24]: currently yes. An update can change, or not change its bookshelf  and/or mark/unmark favorite

File: client/components/BookshelfActions.tsx (hunk 9 above)
Comment (casrom) [reply client/components/BookshelfActions.tsx:40]: might need to add a separate endpoint to get specific userBook info

**Comment by casrom** [comment]
Consider merge to apply front-end changes

**Review by hans-lizihan — APPROVED** [review]

**Pull request merged**